Src/DependencyCollector/Shared/Implementation/SqlClientDiagnosticSourceListener.cs

```diff
@@ -0,0 +1,565 @@
+        {
+            telemetry.Context.Operation.ParentId = operationId.ToString("N");
+        }
+
+        private void OnAfterCommandEvent(
+            Guid operationId,
+            string operation,
+            Guid connectionId,
+            SqlCommand command,
+            IDictionary statistics,
+            long endTimestamp)
+        {
+            // Call DeriveDuration first to ensure we remove any start timestamp from operationStartTimestamps
+            var duration = this.DeriveDuration(operationId, endTimestamp);
+
+            var dependencyName = string.Empty;
+            var target = string.Empty;
+
+            if (command.Connection != null)
+            {
+                target = string.Join(" | ", command.Connection.DataSource, command.Connection.Database);
+
+                var commandName = command.CommandType == CommandType.StoredProcedure
+                    ? command.CommandText
+                    : string.Empty;
+
+                dependencyName = string.IsNullOrEmpty(commandName)
+                    ? string.Join(" | ", command.Connection.DataSource, command.Connection.Database)
+                    : string.Join(" | ", command.Connection.DataSource, command.Connection.Database, commandName);
+            }
+
+            var telemetry = new DependencyTelemetry()
+            {
+                Id = operationId.ToString("N"),
+                Name = dependencyName,
+                Type = RemoteDependencyConstants.SQL,
+                Target = target,
+                Data = command.CommandText,
+                Duration = duration,
+                Timestamp = DateTimeOffset.UtcNow - duration,
+                Success = true
+            };
+
+            InitializeTelemetry(telemetry, operationId);
+            
+            this.client.Track(telemetry);
+        }
+
+        private void OnCommandErrorEvent(
+            Guid operationId,
+            string operation,
+            Guid connectionId,
+            Exception exception,
+            long endTimestamp)
+        {
+            var duration = this.DeriveDuration(operationId, endTimestamp);
+
+            var telemetry = new ExceptionTelemetry(exception)
+            {
+                Message = exception.Message,
+                Timestamp = DateTimeOffset.UtcNow - duration,
+                SeverityLevel = SeverityLevel.Critical
+            };
+
+            InitializeTelemetry(telemetry, operationId);
+            CorrelateErrorTelemetry(telemetry, operationId);
+
+            this.client.Track(telemetry);
+        }
+
+        private void OnAfterConnectionEvent(
+            Guid operationId, 
+            string operation,
+            Guid connectionId,
+            SqlConnection connection, 
+            IDictionary statistics, 
+            long endTimestamp)
+        {
+            var duration = this.DeriveDuration(operationId, endTimestamp);
```

Comment:
how precise are the duration? Does it have StopWatch precision? Open() is typically very fast and we need to ensure our precision can handle such fast calls.